include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1110 @@
+                  H);
+#else
+        AssertThrow(false, ExcNotImplemented());
+#endif
+
+        for (unsigned int i=0; i<n_independent_variables; i++)
+          for (unsigned int j=0; j<i+1; j++)
+            {
+              hessian[i][j] = H[i][j];
+              if (i != j)
+                hessian[j][i] = H[i][j]; // Symmetry
+            }
+
+        // Cleanup :-/
+        for (unsigned int i=0; i<n_independent_variables; i++)
+          delete[] H[i];
+        delete[] H;
+        H = nullptr;
+      }
+
+      // === Vector drivers ===
+
+      static void
+      values (const unsigned int             &active_tape_index,
+              const unsigned int             &n_dependent_variables,
+              const unsigned int             &n_independent_variables,
+              const std::vector<scalar_type> &independent_variables,
+              Vector<scalar_type>            &values)
+      {
+        Assert(independent_variables.size() == n_independent_variables,
+               ExcDimensionMismatch(independent_variables.size(),n_independent_variables));
+        Assert(values.size() == n_dependent_variables,
+               ExcMessage("The length of the dependent function vector must "
+                          " be equal to the number of dependent variables."));
+
+#ifdef DEAL_II_WITH_ADOLC
+        // Note: Adol-C's ::function function expects a *double as the last
+        // parameter. Here we take advantage of the fact that the data in the
+        // Vector class is aligned (e.g. stored as an Array)
+        ::function(active_tape_index,
+                   n_dependent_variables,
+                   n_independent_variables,
+                   const_cast<scalar_type *>(independent_variables.data()),
+                   &values[0]);
+#else
+        AssertThrow(false, ExcNotImplemented());
+#endif
+
+      }
+
+      static void
+      jacobian (const unsigned int             &active_tape_index,
+                const unsigned int             &n_dependent_variables,
+                const unsigned int             &n_independent_variables,
+                const std::vector<scalar_type> &independent_variables,
+                FullMatrix<scalar_type>        &jacobian)
+      {
+        Assert(independent_variables.size() == n_independent_variables,
+               ExcDimensionMismatch(independent_variables.size(),n_independent_variables));
+        Assert(jacobian.m() == n_dependent_variables,
+               ExcMessage("The jacobian row length must be equal "
+                          "to the number of dependent variables."));
+        Assert(jacobian.n() == n_independent_variables,
+               ExcMessage("The hessian column length must be equal "
+                          "to the number of independent variables."));
+
+        scalar_type **J = new scalar_type*[n_dependent_variables];
+        for (unsigned int i=0; i<n_dependent_variables; ++i)
+          J[i] = new scalar_type[n_independent_variables];
```

Comment:
hm, but can you not just use here a pointer to each row in `jacobian` (instead of allocating memory)?

Reply:
i mean `J[i] = &jacobian(i,0)` should work. `FullMatrix` is row-major so this should be equivalent to what you do here.

Reply:
Thanks, the more explicit explanation of what you were inferring was helpful.